[Hashing] Use a non-deterministic seed if LLVM_ENABLE_ABI_BREAKING_CHECKS

MaskRay opened the pull request:
Hashing.h provides hash_value/hash_combine/hash_combine_range, which are
primarily used by `DenseMap<StringRef, X>`

Users shouldn't rely on specific hash values due to size_t differences
on 32-bit/64-bit platforms and potential algorithm changes.
`set_fixed_execution_hash_seed` is provided but it has never been used.

In LLVM_ENABLE_ABI_BREAKING_CHECKS builds, take the the address of a
static storage duration variable as the seed like
absl/hash/internal/hash.h `kSeed`. (See https://reviews.llvm.org/D93931
for workaround for older Clang. Mach-O x86-64 forces PIC, so absl's
`__apple_build_version__` check is unnecessary.)

LLVM_ENABLE_ABI_BREAKING_CHECKS defaults to `WITH_ASSERTS` and is
enabled in an assertion build.

In a non-assertion build, `get_execution_seed` returns the fixed value
regardless of `NDEBUG`. Removing a variable load yields noticeable
size/performance improvement.

A few users relying on the iteration order of `DenseMap<StringRef, X>`
have been fixed (e.g., f8f4235612b9 c025bd1fdbbd 89e8e63f47ff
86eb6bf6715c eb8d03656549 0ea6b8e476c2 58d7a6e0e636 8ea31db27211
592abf29f9f7 664497557ae7).
From my experience fixing [`StringMap`](https://discourse.llvm.org/t/reverse-iteration-bots/72224)
iteration order issues, the scale of issues is similar.

MaskRay committed:
[𝘀𝗽𝗿] changes to main this commit is based on

Created using spr 1.3.5-bogner

[skip ci]

MaskRay committed:
[𝘀𝗽𝗿] initial version

Created using spr 1.3.5-bogner

nikic reviewed (commented):
I like the idea, but I'm not sure this should be always enabled. This sounds like a lot of debugging fun when a seed-dependent case is not caught in testing.

Can we have this part of LLVM_REVERSE_ITERATION instead, which is used to detect similar hash iteration stability issues? We have a build bot for it.

MaskRay commented:
> I like the idea, but I'm not sure this should be always enabled. This sounds like a lot of debugging fun when a seed-dependent case is not caught in testing.
> 
> Can we have this part of LLVM_REVERSE_ITERATION instead, which is used to detect similar hash iteration stability issues? We have a build bot for it.

I have fixed a lot of non-determinism bugs in the past few years.
In my experience, non-determinism caused by `DenseMap<A *, X>` is a more frequent issue than overuse of `DenseMap<StringRef, X>`.
`DenseMap<A *, X>` non-determinism issues could have a low failure rate with certain malloc implementations, especially when the bucket count is small.

Potential non-determinism concerns are outweighed by the benefits.
The argument in https://github.com/abseil/abseil-cpp/issues/339 might be useful:

> Part of Abseil's philosophy is engineering at scale. In addition to making it harder to execute a hash flooding attack, another reason we randomize iteration order is that is makes it easier for use to change the underlying implementation if we need to. When we wanted to change our hash function (for example), we found thousands of unit tests that depended on iteration order. We could not just break these tests and say "sorry, you are violating Hyrum's Law" since we would have thousands of angry Google developers. So instead, we fixed the tests and implemented randomization so that next time we wanted to change something in the implementation, this would not be an issue. The need for iteration order determinism is relatively rare compared to the potential wins we get by being free to change the implementation (maybe to make the hash table faster, which could save millions of cycles at scale, for example). By giving users a knob to disable the randomness, we'd be in the same situation all over again.

While -DLLVM_ENABLE_REVERSE_ITERATION=on helps, it adds complexity for bot maintainers and configurations.
This PR allows easier detection of `DenseMap<StringRef, X>` overuse by all contributors.
Eventually, perhaps we can extend this to all `DenseMap<Key, X>` for LLVM_ENABLE_ASSERTIONS=on.

efriedma-quic commented:
We restricted reverse-iteration when we added it just to save time when we were enabling it: we wanted to prioritize issues that were actually likely to cause non-determinism (as opposed to relying on the hash algorithm, which is annoying but not actually non-deterministic).  If you're willing to fix all the resulting breakage, it should be fine to apply it more widely.

-----

I'm a little concerned that doing this in release builds is going to lead to weird bug reports. Especially given the current approach for getting randomness: ASLR isn't really that random, particularly on Windows, so the probability of getting a particular seed isn't uniform.

MaskRay commented:
Due to ASLR and malloc non-determinism, `DenseMap<A *, X>` already introduces a lot of non-determinism to release builds.
`DenseMap<StringRef, X>` issues are less pronounced compared with `DenseMap<A *, X>`.

> I'm a little concerned that doing this in release builds is going to lead to weird bug reports

I understand the concern. If we receive 10 weird bug reports root caused to `DenseMap<A *, X>`,
the number of `DenseMap<StringRef, X>` bug reports might just be 1.

---

The next step is to make integer type keys (e.g. `DenseMap<uint64_t, X>`) non-deterministic.
IIRC the recent uint64_t getHashKey changes does not require any test update.
The number of `DenseMap<integer, X>` bug reports might be small.

As my previous comment says:

> we could consider restrict the nondeterminism to LLVM_ENABLE_ASSERTIONS=on.

but I haven't found it necessary.

MaskRay committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.5-bogner

[skip ci]

MaskRay committed:
mention size_t differences on  32-bit/64-bit platforms

Created using spr 1.3.5-bogner

MaskRay committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.5-bogner

[skip ci]

MaskRay committed:
non-deterministic if LLVM_ENABLE_ABI_BREAKING_CHECKS

Created using spr 1.3.5-bogner

MaskRay commented:
The non-determinism is now restricted to `LLVM_ENABLE_ABI_BREAKING_CHECKS` builds.

LLVM_ENABLE_ABI_BREAKING_CHECKS defaults to `WITH_ASSERTS` . Release builds that disable assertions disable `LLVM_ENABLE_ABI_BREAKING_CHECKS`. This change yields a slight code size/performance advantage by eliminating a variable read.

nikic commented on llvm/include/llvm/ADT/Hashing.h:
```diff
@@ -322,24 +306,20 @@ struct hash_state {
-  const uint64_t seed_prime = 0xff51afd7ed558ccdULL;
-  static uint64_t seed = fixed_seed_override ? fixed_seed_override : seed_prime;
-  return seed;
+  [[maybe_unused]] static const char seed = 0;
```
Move this inside the `#if`?

nikic reviewed (approved):
LGTM, but please wait a bit in case there is more feedback.

MaskRay committed:
move seed inside #if

Created using spr 1.3.5-bogner

MaskRay commented:
https://llvm-compile-time-tracker.com/compare.php?from=abfff89b743584d2796000318198bf60d3622a1f&to=5c2a6b5ba62d2b7ed2c0ad3be29fba8558f5627b&stat=instructions:u

stage2-O3: `instruction:u` change (larger than expected):

|Benchmark|Old|New|
|--- |--- |--- |
|kimwitu++|38847M|38705M (-0.37%)|
|sqlite3|35002M|34917M (-0.24%)|
|consumer-typeset|31851M|31794M (-0.18%)|
|Bullet|93072M|92815M (-0.28%)|
|tramp3d-v4|78154M|77899M (-0.33%)|
|mafft|32841M|32718M (-0.37%)|
|ClamAV|50244M|50128M (-0.23%)|
|lencod|60998M|60793M (-0.34%)|
|SPASS|42834M|42725M (-0.25%)|
|7zip|191523M|191032M (-0.26%)|
|geomean|55146M|54990M (-0.28%)|

efriedma-quic commented on llvm/include/llvm/ADT/Hashing.h:
```diff
@@ -322,24 +306,20 @@ struct hash_state {
+  // Work around x86-64 negative offset folding for old Clang -fno-pic
+  // https://reviews.llvm.org/D93931
+#if LLVM_ENABLE_ABI_BREAKING_CHECKS &&                                         \
+    (!defined(__clang__) || __clang_major__ > 11)
```
Is it an ABI problem that this ifdef exists?  I mean, LLVM libraries built with clang<11 can't be used by programs built with clang>11.  With LLVM_ENABLE_ABI_BREAKING_CHECKS, I guess it's unlikely to cause issues, though.  (I guess you could use an empty inline asm as a workaround if you wanted to.)

MaskRay replied:
The clang condition introduced a slight ABI problem when mixing `clang<11` and `clang>=11`. In practice it is rare that the llvm-project build and a downstream client exchange the hash values.

joker-eph replied:
Wouldn't this affect any access to a DenseMap across the boundary for example?

That is any API where LLVM would initialize a DenseMap and return a reference to it to the user which then access elements by computing key hashes?

vzakhari replied:
It is definitely a problem for LLVM built using GCC and an external project, based on LLVM, built using CLANG 11.  Is there a reliable check that we can add (e.g. in `llvm/include/llvm/Config/abi-breaking.h.cmake`) to detect this at the build time rather than failing during the execution?

MaskRay replied:
LLVM built using GCC are typically release builds and LLVM_ENABLE_ABI_BREAKING_CHECKS is 0. The `else` branch is taken and there is no impact.

If we want to be over-cautious, this condition can be refined to:
`LLVM_ENABLE_ABI_BREAKING_CHECKS && (!defined(__clang__) || __clang_major__ > 11 || defined(__PIC__))`.
We could code the condition to `utils/bazel/llvm_configs/abi-breaking.h.cmake`, but I am hoping that the complexity isn't needed.

nikic replied:
Maybe it would make sense to move the function to the source file for the LLVM_ENABLE_ABI_BREAKING_CHECKS case? It will add cost, but it's okay for assertion-enabled builds.

vzakhari replied:
In my case it is a `Release` build with GCC and `-DLLVM_ENABLE_ASSERTIONS=ON`.

Could you please explain how `__PIC__` check helps?

chsigg replied:
We are still dealing with the fallout from this change in [Triton](https://github.com/triton-lang/triton/pull/4760).

It seems that breaking the ABI across different compilers is acceptable, but it would be good to detect this and error out. We already [detect mismatches](https://github.com/llvm/llvm-project/blob/bc91f3cdd57cbe4b0a456626f52960158cb3232f/llvm/include/llvm/Config/abi-breaking.h.cmake#L45) of `LLVM_ENABLE_ABI_BREAKING_CHECKS`. Would it be acceptable to just add the `clang > 11` condition there?

efriedma-quic reviewed (commented):
I think I'm happier restricting the non-determinism to +Asserts for now, at least as an incremental step.

> Due to Avalanche effects, even a few ASLR bits are sufficient to cover many different scenarios and expose latent bugs.

On Windows specifically, I'm less concerned about the total number of bits, and more concerned that ASLR isn't randomized for each run of an executable.

MaskRay commented:
Eli is happier now. I plan to land this in a few hours.

MaskRay committed:
[𝘀𝗽𝗿] changes introduced through rebase

Created using spr 1.3.5-bogner

[skip ci]

MaskRay committed:
rebase

Created using spr 1.3.5-bogner

joker-eph commented:
The premerge config seems broken because of this PR.

MaskRay commented:
> The premerge config seems broken because of this PR.
> 
> [lab.llvm.org/buildbot](https://lab.llvm.org/buildbot/#/builders/153/builds/1490)

(The two `lld/test/MachO` failures did not reliably reproduce. Fixed ~3 days ago.)

` BOLT::reader-stale-yaml-std.test` relied on particular hash values, which could not be made reliably passing. Disabled it in  58004e5bb74de607f2f81649fd5a8db3ee318528 and informed BOLT maintainers.

quic-seaswara commented:
Aren't hashes supposed to be deterministic ? 

Can we add a override for builds with asserts turned ON to favor determinism ? 

Could have an API to set the determinism flag ?

This also has an effect on Build ID computation ?

MaskRay commented:
> Aren't hashes supposed to be deterministic ?
> 
> Can we add a override for builds with asserts turned ON to favor determinism ?
> 
> Could have an API to set the determinism flag ?
> 
> This also has an effect on Build ID computation ?

Some hashes are supposed to be deterministic, but `Hashing.h` isn't .
The file-level documentation  and `get_execution_seed` are very clear about the unstability.

If you want to deploy assertions builds and have concern about non-deterministism, set LLVM_ABI_BREAKING_CHECKS to `FORCE_OFF`.

The unstable property is important to ensure that no test relies on the particular implementation and allow us to replace the implementation, which would be infeasible if people keep adding brittle tests.
